### PR TITLE
Add CLI to apply EMOS coefficients

### PR DESCRIPTION
**Pull request by gavinevans**
Addresses #854 

Description
Builds on top of #865 to add a CLI to apply the EMOS coefficients. CLI tests are also added, similar to the existing tests for the ensemble-calibration CLI. The CLI tests show that the ensemble-calibration CLI is equivalent to the estimate-emos-coefficients => apply-emos-coefficients CLIs.  

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/cli/apply_emos_coefficients.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+       Statistics (EMOS), otherwise known as Non-homogeneous Gaussian
+       Regression (NGR). The coefficients are applied to the forecast
+       that is supplied, so as to calibrate the forecast. The calibrated
+       forecast is written to a netCDF file.
+    """
+    parser = ArgParser(
+        description='Apply coefficients for Ensemble Model Output '
+                    'Statistics (EMOS), otherwise known as Non-homogeneous '
+                    'Gaussian Regression (NGR)')
+    # Filepaths for historic and truth data.
```
Comment (bayliffe) [inline lib/improver/cli/apply_emos_coefficients.py]: Nope.

File: tests/improver-apply-emos-coefficients/02-gaussian.bats [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "apply-emos-coefficients for diagnostic with assumed gaussian distribution" {
+  improver_check_skip_acceptance
+  KGO="apply-emos-coefficients/gaussian/kgo.nc"
```
Comment (bayliffe) [inline tests/improver-apply-emos-coefficients/02-gaussian.bats]: Why have a separate KGO to the version in the ensemble-calibration/gaussian/kgo.nc? Are the tests not set up to give the same result? (nccmp-ing the two they appear to be very nearly identical, precision level differences). It would be nice to use one KGO for both ensemble-calibration and this CLI as it would demonstrate equivalence between the one step and two step processes of using ensemble calibration.

File: tests/improver-apply-emos-coefficients/03-trunc_gaussian.bats [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "apply-emos-coefficients for diagnostic with assumed truncated gaussian distribution" {
+  improver_check_skip_acceptance
+  KGO="apply-emos-coefficients/truncated_gaussian/kgo.nc"
```
Comment (bayliffe) [inline tests/improver-apply-emos-coefficients/03-trunc_gaussian.bats]: nccmping here leads to more significant differences. Is this test not equivalent?

File: tests/improver-apply-emos-coefficients/05-current_forecast_probabilities.bats [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "apply-emos-coefficients using probabilities as input" {
+  improver_check_skip_acceptance
+  KGO="apply-emos-coefficients/probabilities/kgo.nc"
```
Comment (bayliffe) [inline tests/improver-apply-emos-coefficients/05-current_forecast_probabilities.bats]: More precision level differences to the single step ensemble-calibration CLI. Is there some float64 step in there which is rounded away in the output  --> input steps of using two CLIs?

File: tests/improver-apply-emos-coefficients/07-current_forecast_percentiles.bats [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "apply-emos-coefficients using percentiles as input" {
+  improver_check_skip_acceptance
+  KGO="apply-emos-coefficients/percentiles/kgo.nc"
```
Comment (bayliffe) [inline tests/improver-apply-emos-coefficients/07-current_forecast_percentiles.bats]: Precision differences to ensemble calibration again.

File: lib/improver/cli/apply_emos_coefficients.py (hunk 1 above)
Comment (gavinevans) [reply lib/improver/cli/apply_emos_coefficients.py]: I've corrected this comment.

File: tests/improver-apply-emos-coefficients/02-gaussian.bats (hunk 2 above)
Comment (gavinevans) [reply tests/improver-apply-emos-coefficients/02-gaussian.bats]: I've updated these CLI tests now so that now there isn't any specific `apply-emos-coefficients` CLI test output, as instead this CLI uses inputs and known good outputs that already exist for the `estimate-emos-coefficients` and `ensemble-calibration` CLIs. This shows the equivalency of either running the `ensemble-calibration` CLI that estimates the coefficients and applies them within one CLI versus the combination of the `estimate-emos-coefficients` and `apply-emos-coefficients`.

File: tests/improver-apply-emos-coefficients/03-trunc_gaussian.bats (hunk 3 above)
Comment (gavinevans) [reply tests/improver-apply-emos-coefficients/03-trunc_gaussian.bats]: This should no longer exist as the known good output from the `apply-emos-coefficients` CLI is the output from the `ensemble-calibration` CLI.

File: tests/improver-apply-emos-coefficients/05-current_forecast_probabilities.bats (hunk 4 above)
Comment (gavinevans) [reply tests/improver-apply-emos-coefficients/05-current_forecast_probabilities.bats]: This is now rectified.

File: tests/improver-apply-emos-coefficients/07-current_forecast_percentiles.bats (hunk 5 above)
Comment (gavinevans) [reply tests/improver-apply-emos-coefficients/07-current_forecast_percentiles.bats]: This has been resolved.

File: tests/improver-apply-emos-coefficients/02-gaussian.bats (hunk 2 above)
Comment (bayliffe) [reply tests/improver-apply-emos-coefficients/02-gaussian.bats]: Thanks.

**Comment by bayliffe** [comment]
Thanks for the changes Gavin, particularly reconciling the tests between this approach and the single CLI to calculate and apply coefficients approach. There's the one question about the CLI data you've plotted but I expect the explanation to be trivial. so I am happy that this is ready to go.

File: lib/improver/cli/apply_emos_coefficients.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                        help='A path to an input NetCDF file containing the '
+                             'forecast to be calibrated.')
+    parser.add_argument('coefficients_filepath',
+                        metavar='COEFFICIENTS_FILEPATH',
+                        help='A path to an input NetCDF file containing the '
+                             'coefficients used for calibration.')
+    parser.add_argument('output_filepath', metavar='OUTPUT_FILEPATH',
+                        help='The output path for the processed NetCDF')
+    # Optional arguments.
+    parser.add_argument('--num_realizations', metavar='NUMBER_OF_REALIZATIONS',
```
Comment (fionaRust) [inline lib/improver/cli/apply_emos_coefficients.py]: What happens with the default now, if there are probabilities or percentiles as input. Does this need to be a compulsory argument now? Or at least saying it will raise an error if not provided

File: lib/improver/cli/apply_emos_coefficients.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                             'If the current forecast is input as '
+                             'probabilities or percentiles then this argument '
+                             'is used to create the requested number of '
+                             'realizations. In addition, this argument is '
+                             'used to construct the requested number of '
+                             'realizations from the mean and variance output '
+                             'after applying the EMOS coefficients.'
+                             'Default will be the number of realizations '
+                             'in the raw input file.')
+    parser.add_argument('--random_ordering', default=False,
```
Comment (fionaRust) [inline lib/improver/cli/apply_emos_coefficients.py]: Might be worth saying this is only used for realization input.

File: lib/improver/cli/apply_emos_coefficients.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+       Regression (NGR). The coefficients are applied to the forecast
+       that is supplied, so as to calibrate the forecast. The calibrated
+       forecast is written to a netCDF file.
+    """
+    parser = ArgParser(
+        description='Apply coefficients for Ensemble Model Output '
+                    'Statistics (EMOS), otherwise known as Non-homogeneous '
+                    'Gaussian Regression (NGR)')
+    # Filepaths for the forecast, EMOS coefficients and the output.
+    parser.add_argument('forecast_filepath', metavar='FORECAST_FILEPATH',
```
Comment (fionaRust) [inline lib/improver/cli/apply_emos_coefficients.py]: Need to say this can be in percentiles, probabilities or realizations and that the output will be in the same form. In general the description could maybe have a bit more information about the steps the different input forms will go through. What do you think?

File: tests/improver-apply-emos-coefficients/09-current_forecast_rebadged_percentiles.bats [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  # Run nccmp to compare the output calibrated realizations when the input
+  # is percentiles that have been rebadged as realizations. The known good
+  # output in this case is the same as when passing in percentiles directly,
+  # apart from a difference in the coordinates, such that the percentile input
+  # will have a percentile coordinate, whilst the rebadged percentile input
+  # will result in a realization coordinate.
+  # The "--exclude=realization" option indicates that the realization coordinate
+  # should be ignored.
+  # The "--exclude=percentile" option indicates that the percentile coordinate
+  # should be ignored.
```
Comment (fionaRust) [inline tests/improver-apply-emos-coefficients/09-current_forecast_rebadged_percentiles.bats]: Could also add that the -t option matches the precision in the  improver_compare_output_lower_precision function used elsewhere.

File: tests/improver-ensemble-calibration/07-current_forecast_percentiles.bats [hunk 10]
```diff
@@ -40,7 +40,7 @@
       'gaussian' "$IMPROVER_ACC_TEST_DIR/ensemble-calibration/percentiles/input.nc" \
       "$IMPROVER_ACC_TEST_DIR/ensemble-calibration/gaussian/history/*.nc" \
       "$IMPROVER_ACC_TEST_DIR/ensemble-calibration/gaussian/truth/*.nc" \
-      "$TEST_DIR/output.nc" --num_realizations=18 --max_iterations 200
```
Comment (fionaRust) [inline tests/improver-ensemble-calibration/07-current_forecast_percentiles.bats]: Why this change?

File: lib/improver/cli/apply_emos_coefficients.py (hunk 6 above)
Comment (gavinevans) [reply lib/improver/cli/apply_emos_coefficients.py]: I've amended this docstring, as discussed.

File: lib/improver/cli/apply_emos_coefficients.py (hunk 7 above)
Comment (gavinevans) [reply lib/improver/cli/apply_emos_coefficients.py]: Extended docstring.

File: lib/improver/cli/apply_emos_coefficients.py (hunk 8 above)
Comment (gavinevans) [reply lib/improver/cli/apply_emos_coefficients.py]: I've added a bit more information here.

File: tests/improver-apply-emos-coefficients/09-current_forecast_rebadged_percentiles.bats (hunk 9 above)
Comment (gavinevans) [reply tests/improver-apply-emos-coefficients/09-current_forecast_rebadged_percentiles.bats]: Added.

File: tests/improver-ensemble-calibration/07-current_forecast_percentiles.bats (hunk 10 above)
Comment (gavinevans) [reply tests/improver-ensemble-calibration/07-current_forecast_percentiles.bats]: The `max_iterations` argument was previously included for stability reasons, however, is no longer required.
